### PR TITLE
feat(swarm): enforce creation of `Swarm` via `SwarmBuilder`

### LINKED ISSUE #3107: swarm: Deprecate `futures::executor::ThreadPool` usage for connection tasks
## Description

Currently, the default executor for connection tasks is a dedicated thread-pool created within the `SwarmBuilder`. With https://github.com/libp2p/rust-libp2p/pull/3097, this is being made more explicit.

As a follow-up we should deprecate the use of `ThreadPool` entirely.

## Motivation

Users of `rust-libp2p` will need to use some kind of async executor. They should use one of the new executor APIs introduced in https://github.com/libp2p/rust-libp2p/pull/3097.

## Are you planning to do it yourself in a pull request?

Maybe.

### LINKED ISSUE #3186: swarm: Deprecate `Swarm::with_XYZ` and enforce creation via `SwarmBuilder`
## Description

With the introduction of the `Swarm::with_executor` APIs (#3068), we have a fair bit of duplication in APIs on how to create a `Swarm`. Additionally, the current `SwarmBuilder` and `Swarm` APIs don't work well with the QUIC and WebRTC transports as those don't go through the typical builder pattern on `Transport` and hereby don't construct a `StreamMuxerBox` (see https://github.com/libp2p/rust-libp2p/issues/3179).

We should streamline the creation of a `Swarm` to work with any `Transport` and be easy to discover. @mxinden [suggested](https://github.com/libp2p/rust-libp2p/issues/3186#issuecomment-1342845258) to achieve this by forcing users to always go through `SwarmBuilder`.

## Motivation

- Less API surface and layers of indirection.

## Are you planning to do it yourself in a pull request?

Maybe.

### PR DESCRIPTION
**Pull request by vnermolaev**
## Description

Mark constructors `Swarm::with_X_executor` as deprecated.
Move the deprecated functionality to `SwarmBuilder::with_X_executor`
Use `SwarmBuilder` throughout.

Resolves #3186.
Resolves #3107.

## Notes
The main file to audit is https://github.com/libp2p/rust-libp2p/pull/3588/files#diff-7d75430c06d5d5aeba4229b6535d2b56f4e67809ffdac99d2dec8eb8a81b3c5c.
The rest contain changes related to the API change.

## Change checklist

Changelog is to be updated after the tech side is settled.

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] A changelog entry has been made in the appropriate crates

### CODE REVIEW

File: swarm/src/lib.rs [hunk 1]
```diff
@@ -1508,24 +1527,79 @@ impl<TBehaviour> SwarmBuilder<TBehaviour>
 where
     TBehaviour: NetworkBehaviour,
 {
-    /// Creates a new [`SwarmBuilder`] from the given transport, behaviour, local peer ID and
-    /// executor. The `Swarm` with its underlying `Network` is obtained via
-    /// [`SwarmBuilder::build`].
-    pub fn with_executor(
```
Comment (thomaseizinger) [inline swarm/src/lib.rs]: This diff makes it look like this function was removed when in fact it is now on line 1594. Can we restore the order please?

File: swarm/src/lib.rs [hunk 2]
```diff
@@ -1508,24 +1527,79 @@ impl<TBehaviour> SwarmBuilder<TBehaviour>
...
             connection_limits: Default::default(),
         }
     }
 
+    /// Private convenience function.
+    /// Sets executor of this [`SwarmBuilder`] to a given value.
+    fn executor(mut self, executor: impl Executor + Send + 'static) -> Self {
+        self.pool_config = PoolConfig::new(Some(Box::new(executor)));
+        self
+    }
```
Comment (thomaseizinger) [inline swarm/src/lib.rs]: Do we need this? There shouldn't be a need to set the executor past constructing the builder.

File: swarm/src/lib.rs [hunk 3]
```diff
@@ -1537,12 +1611,7 @@ where
         behaviour: TBehaviour,
         local_peer_id: PeerId,
     ) -> Self {
-        Self::with_executor(
-            transport,
-            behaviour,
-            local_peer_id,
-            crate::executor::TokioExecutor,
-        )
+        Self::new(transport, behaviour, local_peer_id).executor(crate::executor::TokioExecutor)
```
Comment (thomaseizinger) [inline swarm/src/lib.rs]: Is this diff necessary?

File: swarm/src/lib.rs [hunk 4]
```diff
@@ -349,11 +349,15 @@ where
...
 impl<TBehaviour> Unpin for Swarm<TBehaviour> where TBehaviour: NetworkBehaviour {}
 
+#[allow(deprecated)]
 impl<TBehaviour> Swarm<TBehaviour>
 where
     TBehaviour: NetworkBehaviour,
 {
     /// Builds a new `Swarm` with a provided executor.
+    #[deprecated(
+        note = "Use a builder chain `SwarmBuilder::with_executor(transport, behaviour, local_peer_id, executor).build()` instead."
```
Comment (thomaseizinger) [inline swarm/src/lib.rs]: I think we can shorten this to:

```suggestion
        note = "Use `SwarmBuilder::with_executor` instead."
```

**Review by thomaseizinger — COMMENTED** [review]
Thank you for this! Left some comments.

File: swarm/src/lib.rs (hunk 2 above)
Comment (vnermolaev) [reply swarm/src/lib.rs]: Strictly speaking, we don't. This is a private convenience function; the end user will never see it. I can ofc do the construction in place, but this will lead to a lot of code duplication.

Comment (thomaseizinger) [reply swarm/src/lib.rs]: Unless I am missing something, this field should only be set in two places:

1. A constructor that sets an executor
2. A constructor that doesn't set an executor

Shouldn't it be possible to implement everything else on top of these? Two (different) usages of the same field isn't critical code duplication to me.

File: swarm/src/lib.rs (hunk 4 above)
Comment (vnermolaev) [reply swarm/src/lib.rs]: I will replace it, however, the message I propose could be used as a direct replacement.

Comment (thomaseizinger) [reply swarm/src/lib.rs]: > I will replace it, however, the message I propose could be used as a direct replacement.

That only works if users use these exact variable names. The only thing we can say for sure is the name of the function so that is what I'd like to point to.

The story would be different if Rust had support for codemods :)

File: swarm/src/lib.rs (hunk 2 above)
Comment (vnermolaev) [reply swarm/src/lib.rs]: fixed

File: swarm/src/lib.rs [hunk 5]
```diff
@@ -1527,6 +1534,47 @@ where
...
+            .name_prefix("libp2p-swarm-task-")
+            .create()
+        {
+            Ok(tp) => Self::with_executor(transport, behaviour, local_peer_id, tp),
+            Err(err) => {
+                log::warn!("Failed to create executor thread pool: {:?}", err);
+                Self::without_executor(transport, behaviour, local_peer_id)
+            }
+        }
+    }
```
Comment (thomaseizinger) [inline swarm/src/lib.rs]: Let's not introduce this one. See https://github.com/libp2p/rust-libp2p/issues/3107.

File: swarm/src/lib.rs [hunk 6]
```diff
@@ -401,6 +405,7 @@ where
     }
 
     /// Builds a new `Swarm` with a threadpool executor.
+    #[deprecated(note = "Use `SwarmBuilder::with_threadpool_executor` instead.")]
```
Comment (thomaseizinger) [inline swarm/src/lib.rs]:
```suggestion
    #[deprecated(note = "The `futures::executor::ThreadPool` executor is deprecated. See https://github.com/libp2p/rust-libp2p/issues/3107.")]
```

**Review by thomaseizinger — COMMENTED** [review]
Nice, thank you for that!

Can you add a changelog entry to `libp2p-swarm` please?

File: swarm/src/lib.rs (hunk 5 above)
Comment (vnermolaev) [reply swarm/src/lib.rs]: Actually, I cannot simply not introduce it. 
```bash
error[E0599]: no function or associated item named `with_threadpool_executor` found for struct `SwarmBuilder` in the current scope
  --> examples/file-sharing/src/network.rs:50:31
   |
50 |     let swarm = SwarmBuilder::with_threadpool_executor(
   |                               ^^^^^^^^^^^^^^^^^^^^^^^^
   |                               |
   |                               function or associated item not found in `SwarmBuilder<_>`
   |                               help: there is an associated function with a similar name: `with_executor`
```
So I keep it for now and then it will be removed as part of #3107?

File: swarm/CHANGELOG.md [hunk 7]
```diff
@@ -87,6 +87,9 @@
 
 - Remove `ConnectionId::new`. Manually creating `ConnectionId`s is now unsupported. See [PR 3327].
 
+- Deprecate methods `Swarm::with_executor`, `Swarm::with_*_executor`, `Swarm::without_executor`.
+  Introduce similar methods in `SwarmBuilder`. See [PR3588].
```
Comment (thomaseizinger) [inline swarm/CHANGELOG.md]:
```suggestion
  Introduce similar methods in `SwarmBuilder`. See [PR 3588].
```

File: swarm/CHANGELOG.md [hunk 8]
```diff
@@ -102,6 +105,7 @@
 [PR 3375]: https://github.com/libp2p/rust-libp2p/pull/3375
 [PR 3254]: https://github.com/libp2p/rust-libp2p/pull/3254
 [PR 3497]: https://github.com/libp2p/rust-libp2p/pull/3497
+[PR 3497]: https://github.com/libp2p/rust-libp2p/pull/3588
```
Comment (thomaseizinger) [inline swarm/CHANGELOG.md]:
```suggestion
[PR 3588]: https://github.com/libp2p/rust-libp2p/pull/3588
```

File: swarm/src/lib.rs (hunk 5 above)
Comment (thomaseizinger) [reply swarm/src/lib.rs]: I'd like to not introduce code only to deprecate it later again. Would you mind porting the file-sharing example to a different executor?

**Review by thomaseizinger — APPROVED** [review]
Excellent, thank you!
